File: RubyIrrgarten/Irrgarten1/Labyrinth.rb
require_relative 'Monster'
require_relative 'Player'
require_relative 'Orientation'
require_relative 'Directions'
require_relative 'Dice'

module Irrgarten
  class Labyrinth
    @@BLOCK_CHAR = 'X'
    @@EMPTY_CHAR = '-'
    @@MONSTER_CHAR = 'M'
    @@COMBAT_CHAR = 'C'
    @@EXIT_CHAR = 'E'
    @@ROW = 0
    @@COL = 1

    def initialize(n_rows, n_cols,exit_row,exit_col)
      @n_rows = n_rows
      @n_cols = n_cols
      @exit_row = exit_row
      @exit_col = exit_col
      @monsters = Array.new(n_rows){Array.new(n_cols)}
      @players = Array.new(n_rows){Array.new(n_cols)}
      @labyrinth = Array.new(n_rows){Array.new(n_cols,@@EMPTY_CHAR)}
      @labyrinth[exit_row][exit_col] = @@EXIT_CHAR
    end

    def spread_players(players)
      for i in 0...players.length
        p = players[i]
        pos = random_empty_pos
        put_player_2d(-1,-1,pos[@@ROW],pos[@@COL],p)
      end
    end

    def have_a_winner
      return @players[@exit_row][@exit_col] != nil
    end

    def to_s
      devolver = ""
      for i in 0...@n_rows
        devolver += "| "
        for j in 0...@n_cols
          devolver += @labyrinth[i][j].to_s + " "
        end
        devolver += " |\n"
      end

      return devolver

    end

    def add_monster(row,col,monster)
        if pos_ok(row,col) && @labyrinth[row][col] == @@EMPTY_CHAR
          @labyrinth[row][col] = @@MONSTER_CHAR
          @monsters[row][col] = monster
          monster.set_pos(row,col)
          add = false
        end

    end

    def put_player(direction,player)
      old_row = player.get_row
      old_col = player.get_col
      new_pos = dir2_pos(old_row,old_col,direction)
      monster = put_player_2d(old_row,old_col,new_pos[@@ROW],new_pos[@@COL],player)
      return monster
    end

    def add_block(orientation,start_row,start_col,length)
      if orientation == Orientation::VERTICAL
        inc_row = 1
        inc_col = 0
      else
        inc_row = 0
        inc_col = 1
      end
      row = start_row
      col = start_col

      while pos_ok(row,col) && empty_pos(row,col) && length > 0
        @labyrinth[row][col] = @@BLOCK_CHAR
        length -=1
        row += inc_row
        col += inc_col
      end
    end

    def valid_moves(row,col)
      output = Array.new
      if can_step_on(row+1,col)
        output.push(Directions::DOWN)
      end
      if can_step_on(row-1,col)
        output.push(Directions::UP)
      end
      if can_step_on(row,col+1)
        output.push(Directions::RIGHT)
      end
      if can_step_on(row,col-1)
        output.push(Directions::LEFT)
      end
      return output
    end

    def empty_pos(row,col)
      return @labyrinth[row][col] == @@EMPTY_CHAR
    end
    private

    def pos_ok(row,col)
      return row < @n_rows && col < @n_cols && row >= 0 && col >= 0
    end

    def monster_pos(row,col)
      return @labyrinth[row][col] == @@MONSTER_CHAR
    end

    def exit_pos(row,col)
      return @labyrinth[row][col] == @@EXIT_CHAR
    end

    def combat_pos(row,col)
      return @labyrinth[row][col] == @@COMBAT_CHAR
    end

    def can_step_on(row,col)
      return pos_ok(row,col) && (monster_pos(row,col) || empty_pos(row,col) || exit_pos(row,col))
    end

    def update_old_pos(row,col)
      if pos_ok(row,col)
        if combat_pos(row,col)
          @labyrinth[row][col] = @@MONSTER_CHAR
        else
          @labyrinth[row][col] = @@EMPTY_CHAR
        end
      end
    end

    def dir2_pos(row,col,direction)
      devolver = Array.new
      case direction
      when 'left'
        col -= 1
      when 'right'
        col += 1
      when 'up'
        row += 1
      else
        row -= 1
      end
      devolver.push(row,col)
      return devolver
    end

    def random_empty_pos
      devolver = Array.new
      begin
        num1 = Dice.random_pos(@n_rows)
        num2 = Dice.random_pos(@n_cols)
      end while empty_pos(num1,num2)
      devolver.push(num1,num2)
      return devolver
    end

    def put_player_2d(old_row,old_col,row,col,player)
      output = nil
      if can_step_on(row,col)
        if pos_ok(old_row,old_col)
          p = @players[old_row][old_col]
          if p == player
            update_old_pos(old_row,old_col)
            @players[old_row][old_col] = nil
          end
        end
        monster_pos = monster_pos(row,col)
        if monster_pos
          @labyrinth[row][col] = @@COMBAT_CHAR
          output = @monsters[row][col]
        else
          number = player.get_number
          @labyrinth[row][col] = number
        end
        @players[row][col] = player
        player.set_pos(row,col)
      end
      return output
    end
  end
end

File: src/irrgarten/Labyrinth.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */

package irrgarten;




/**
 *
 * @author aulas
 */
public class Labyrinth {
    private static char BLOCK_CHAR = 'X';
    private static char EMPTY_CHAR = '-';
    private static char MONSTER_CHAR = 'M';
    private static char COMBAT_CHAR = 'C';
    private static char EXIT_CHAR = 'E';
    private static final int ROW = 0;
    private static final int COL = 1;
    
    private int nRows ;
    private int nCols;
    private int exitRow;
    private int exitCol;
    private Player[][] players;
    private char[][] labyrinth;
    private Monster[][] monsters;
    
    private static final int NUM_VECTOR = 2;
    
    public Labyrinth(int nRows,int nCols,int exitRow,int exitCol){
        players = new Player[nRows][nCols];
        monsters = new Monster[nRows][nCols];
        labyrinth = new char[nRows][nCols];
        
        for(int i=0;i<nRows;i++){
            for(int j=0;j<nCols;j++){
                labyrinth[i][j] = EMPTY_CHAR;
            }
        }
        this.nRows = nRows;
        this.nCols = nCols;
        this.exitCol = exitCol;
        this.exitRow = exitRow;
        labyrinth[exitRow][exitCol] = EXIT_CHAR;
    }
    
    public void spreadPlayers(Player[] players){
        for (Player p : players) {          
            int[] pos = randomEmptyPos();
            putPlayer2D(-1,-1,pos[ROW],pos[COL],p);
        }
    }
    
    public boolean haveAWinner(){
        return players[exitRow][exitCol] != null;
    }
    
    @Override
    public String toString(){
        String devolver="";
        for(int i=0;i<nRows;i++){
            devolver += "| ";
            for(int j=0;j<nCols;j++){
                devolver += labyrinth[i][j] +" " ;
            }
        devolver += " |\n";
        }
        return devolver;
    }
    
    public void addMonster(int row,int col,Monster monster){boolean seguir = true;
        int veces = 0;
        while(seguir && veces < 5){
            if(posOK(row,col) && labyrinth[row][col] == EMPTY_CHAR){
                monsters[row][col] = monster;
                monster.setPos(row, col);
                labyrinth[row][col] = 'M';}
            veces++;
        }
    }
    
    public Monster putPlayer(Directions direction,Player player){
       int oldRow = player.getRow();
       int oldCol = player.getCol();
       int[] newPos = dir2Pos(oldRow,oldCol,direction);
       Monster monster = putPlayer2D(oldRow,oldCol,newPos[ROW],newPos[COL],player);
       return monster;
    }
    
    public void addBlock(Orientation orientation,int startRow,int startCol,int length){//este comprobado
        int incRow,incCol;
        
        if(orientation == Orientation.VERTICAL){
            incRow = 1;
            incCol = 0;
        }else{
            incRow = 0;
            incCol = 1;
        }
        
        int row = startRow;
        int col = startCol;
        
        while((posOK(row,col)&& emptyPos(row,col)) && (length > 0)){
            labyrinth[row][col] = BLOCK_CHAR;
            length -= 1;
            row += incRow;
            col += incCol;
        }
        
    }
    
    public void setPlayer(Player p){
        labyrinth[p.getRow()][p.getCol()] = p.getNumber();
        players[p.getRow()][p.getCol()] = p;
    }
    
    public Directions[] validMoves(int row,int col){//comprobado
        int counter = 0;
        Directions[] output = new Directions[4];
        if(canStepOn(row+1,col)){
            output[counter++] = (Directions.DOWN);
        }
        if(canStepOn(row-1,col)){
            output[counter++] = Directions.UP;
        }
        if(canStepOn(row,col+1)){
            output[counter++] = Directions.RIGHT;
        }
        if(canStepOn(row,col-1)){
            output[counter++] = Directions.LEFT;
        }
        return output;
    }
    
    private boolean posOK(int row,int col){
        return row < nRows && col < nCols && row >= 0 && col >= 0;
    }
    
    private boolean emptyPos(int row,int col){
        return labyrinth[row][col] == EMPTY_CHAR;
    }
    
    private boolean monsterPos(int row,int col){
        
        return  labyrinth[row][col] == MONSTER_CHAR;
    }
    
    private boolean exitPos(int row,int col){
        return labyrinth[row][col] == EXIT_CHAR;
    }
    
    private boolean combatPos(int row,int col){
        return labyrinth[row][col] == COMBAT_CHAR;
    }
    private boolean canStepOn(int row,int col){
        return posOK(row,col) && (emptyPos(row,col) || monsterPos(row,col) || exitPos(row,col));
    }
    
    private void updateOldPos(int row,int col){
        if(posOK(row,col)){
            if(combatPos(row,col)){
                labyrinth[row][col] = MONSTER_CHAR;
            }else{
                labyrinth[row][col] = EMPTY_CHAR;
            }
        }
    }
    
    private int[] dir2Pos(int row,int col,Directions direction){
        int[] almacena = new int[NUM_VECTOR];
        switch(direction){
            case UP:
                if(labyrinth[row-1][col] != BLOCK_CHAR){
                    row--;
                }
                break;
            case DOWN:
                if(labyrinth[row+1][col] != BLOCK_CHAR){
                    row++;
                }
                break;
            case RIGHT:
                if(labyrinth[row][col+1] != BLOCK_CHAR){
                    col++;
                }
                break;
            default:
                if(labyrinth[row][col-1] != BLOCK_CHAR){
                    col--;
                }
        }
        
        almacena[0] = row;
        almacena[1] = col;
        return almacena;
    }
    
    private int[] randomEmptyPos(){
        int[] almacena  = new int[NUM_VECTOR];
        boolean blocked = true;
        int num1=0,num2=0;
        while(blocked){
            num1 = Dice.randomPos(nRows);
            num2 = Dice.randomPos(nCols);
            if(labyrinth[num1][num2] != BLOCK_CHAR){
                blocked = false;
            }
        }
        almacena[0] = num1;
        almacena[1] = num2;
        return almacena;
    }
    
    private Monster putPlayer2D(int oldRow,int oldCol,int row,int col,Player player){
         Monster output = null;
         if(canStepOn(row,col)){
             if(posOK(oldRow,oldCol)){
                 Player p = players[oldRow][oldCol];
                 if(p==player){
                     updateOldPos(oldRow,oldCol);
                     players[oldRow][oldCol] = null;
                     
                 }
             }
             
             boolean monsterPos = monsterPos(row,col);
             if(monsterPos){
                 labyrinth[row][col] = COMBAT_CHAR;
                 output = monsters[row][col];
             }else{
                 char number = player.getNumber();
                 labyrinth[row][col] = number;
             }
             
             players[row][col] = player;
             player.setPos(row,col);
         }
         return output;
    }
}
